src/speculators/train/eagle3/core.py

```diff
@@ -0,0 +1,339 @@
+    # shape: [1, total_seq_len - ttt_step]
+
+    correct = predicted_tokens == target_tokens
+    cond_denom: torch.Tensor | int = correct.numel()
+    if prev_correct is not None:
+        cond_denom = prev_correct.sum()
+        # Update prev_correct in place
+        correct = torch.logical_and(prev_correct, correct, out=prev_correct)
+    if loss_mask is not None:
+        correct = torch.masked_select(correct, loss_mask.to(torch.bool))
+
+    correct_sum = correct.float().sum()
+    full_denom = correct.numel()
+
+    return correct_sum / (full_denom + 1e-5), correct_sum / (cond_denom + 1e-5)
+
+
+def loss_function(
+    logits: torch.Tensor,  # shape: [1, total_seq_len - ttt_step, draft_vocab_size]
+    targets: torch.Tensor,  # shape: [1, total_seq_len - ttt_step, draft_vocab_size]
+    loss_mask: torch.Tensor | None,  # shape: [1, total_seq_len - ttt_step]
+):
+    # Note: logits, targets, and loss_mask are already aligned for the current ttt_step
+    logits = torch.nn.functional.log_softmax(logits, dim=-1)
+    target_p = torch.nn.functional.softmax(targets, dim=-1)
+    elementwise_loss = torch.nn.functional.kl_div(
+        logits, target_p, reduction="none", log_target=False
+    )
+
+    if loss_mask is not None:
+        elementwise_loss = elementwise_loss * loss_mask.unsqueeze(-1)
+        denominator: torch.Tensor | int = loss_mask.sum(dim=1) + 1e-5
+    else:
+        denominator = logits.shape[1]  # total_seq_len - ttt_step
+    batch_loss = torch.sum(elementwise_loss, dim=(1, 2)) / denominator
+    # shape: [1]
+    return batch_loss.mean()
+
+
+@SpeculatorModel.register("eagle3_draft")
+class Eagle3DraftModel(SpeculatorModel):
+    config_class: ClassVar[type[Eagle3SpeculatorConfig]] = Eagle3SpeculatorConfig  # type: ignore[misc]
+    _keys_to_ignore_on_load_missing: ClassVar[list[str]] = [  # type: ignore[misc]
+        "embed_tokens.weight",
+        "verifier_lm_head.weight",
+        "d2t",
+        "t2d",
+    ]
+    _keys_to_ignore_on_save: ClassVar[list[str]] = ["verifier_lm_head.weight"]  # type: ignore[misc,assignment]
+
+    def __init__(
+        self,
+        config: Eagle3SpeculatorConfig,
+        t2d: torch.Tensor,
+        d2t: torch.Tensor,
+        ttt_steps: int = 3,
+        ttt_step_loss_decay: float = 1.0,
+    ):
+        super().__init__(
+            config=config,
+            verifier=None,
+            verifier_attachment_mode="train_only",
+        )
+        self.hidden_size = config.transformer_layer_config.hidden_size
+        self.num_layers = config.transformer_layer_config.num_hidden_layers
+        self.decoder_layer_config = config.transformer_layer_config
+        self.ttt_steps = ttt_steps
+        self.ttt_step_loss_decay = ttt_step_loss_decay
+        self.register_buffer("t2d", t2d)  # shape: [verifier_vocab_size], bool
+        self.register_buffer("d2t", d2t)  # shape: [draft_vocab_size], int offsets
+        self.draft_vocab_size = int(t2d.sum(dtype=torch.long).item())
```

Comment:
probably want to assert shape  of t2d/d2t but take the draft vocab size from the config